# ADS 509 Module 1: APIs and Web Scraping

This notebook has three parts. In the first part you will pull data from the Twitter API. In the second, you will scrape lyrics from AZLyrics.com. In the last part, you'll run code that verifies the completeness of your data pull. 

For this assignment you have chosen two musical artists who have at least 100,000 Twitter followers and 20 songs with lyrics on AZLyrics.com. In this part of the assignment we pull the some of the user information for the followers of your artist and store them in text files. 


## Important Note

This assignment requires you to have a version of Tweepy that is at least version 4. The latest version is 4.10 as I write this. Critically, this version of Tweepy is *not* on the upgrade path from Version 3, so you will not be able to simply upgrade the package if you are on Version 3. Instead you will need to explicitly install version 4, which you can do with a command like this: `pip install "tweepy>=4"`. You will also be using Version 2 of the Twitter API for this assignment. 

Run the below cell. If your version of Tweepy begins with a "4", then you should be good to go. If it begins with a "3" then run the following command, found [here](https://stackoverflow.com/questions/5226311/installing-specific-package-version-with-pip), at the command line or in a cell: `pip install -Iv tweepy==4.9`. (You may want to update that version number if Tweepy has moved on past 4.9. 

In [1]:
pip show tweepy

Name: tweepy
Version: 4.12.1
Summary: Twitter library for Python
Home-page: https://www.tweepy.org/
Author: Joshua Roesslein
Author-email: tweepy@googlegroups.com
License: MIT
Location: C:\Users\zfreitas\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages
Requires: oauthlib, requests, requests-oauthlib
Required-by: 
Note: you may need to restart the kernel to use updated packages.


## General Assignment Instructions

These instructions are included in every assignment, to remind you of the coding standards for the class. Feel free to delete this cell after reading it. 

One sign of mature code is conforming to a style guide. We recommend the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html). If you use a different style guide, please include a cell with a link. 

Your code should be relatively easy-to-read, sensibly commented, and clean. Writing code is a messy process, so please be sure to edit your final submission. Remove any cells that are not needed or parts of cells that contain unnecessary code. Remove inessential `import` statements and make sure that all such statements are moved into the designated cell. 

Make use of non-code cells for written commentary. These cells should be grammatical and clearly written. In some of these cells you will have questions to answer. The questions will be marked by a "Q:" and will have a corresponding "A:" spot for you. *Make sure to answer every question marked with a `Q:` for full credit.* 


# Twitter API Pull

In [2]:
# for the twitter section
import tweepy
import os
import datetime
import re
from pprint import pprint

# for the lyrics scrape section
import requests
import time
from bs4 import BeautifulSoup
from collections import defaultdict, Counter

In [3]:
# Use this cell for any import statements you add
# This is a PEP8 linter for my Jupyter Notebooks code.
%load_ext pycodestyle_magic


We need bring in our API keys. Since API keys should be kept secret, we'll keep them in a file called `api_keys.py`. This file should be stored in the directory where you store this notebook. The example file is provided for you on Blackboard. The example has API keys that are _not_ functional, so you'll need to get Twitter credentials and replace the placeholder keys. 

In [4]:
from api_keys import api_key, api_key_secret, bearer_token

In [5]:
# Use the wait_on_rate_limit argument if you don't handle the exception yourself
client = tweepy.Client(bearer_token,wait_on_rate_limit=True)

# Testing the API

The Twitter APIs are quite rich. Let's play around with some of the features before we dive into this section of the assignment. For our testing, it's convenient to have a small data set to play with. We will seed the code with the handle of John Chandler, one of the instructors in this course. His handle is `@37chandler`. Feel free to use a different handle if you would like to look at someone else's data. 

We will write code to explore a few aspects of the API: 

1. Pull some of the followers @37chandler.
1. Explore response data, which gives us information about Twitter users. 
1. Pull the last few tweets by @37chandler.


In [6]:
handle = "37chandler"
user_obj = client.get_user(username=handle)

followers = client.get_users_followers(
    # Learn about user fields here: 
    # https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/user
    user_obj.data.id, user_fields=["created_at","description","location",
                                   "public_metrics"]
)



Now let's explore these a bit. We'll start by printing out names, locations, following count, and followers count for these users. 

In [7]:
num_to_print = 20

for idx, user in enumerate(followers.data) :
    following_count = user.public_metrics['following_count']
    followers_count = user.public_metrics['followers_count']
    
    print(f"{user.name} lists '{user.location}' as their location.")
    print(f" Following: {following_count}, Followers: {followers_count}.")
    print()
    
    if idx >= (num_to_print - 1) :
        break
    

John chandler lists 'Decatur, GA' as their location.
 Following: 130, Followers: 10.

Frank P Seidl lists 'Twin Cities, Minnesota USA' as their location.
 Following: 37865, Followers: 37561.

Roberta lists 'Salinas' as their location.
 Following: 1894, Followers: 181.

Anna bikes MKE lists 'mke ' as their location.
 Following: 2287, Followers: 1756.

Catherine lists 'San Angelo' as their location.
 Following: 2189, Followers: 224.

Lisa lists 'None' as their location.
 Following: 2596, Followers: 831.

Lexi lists 'None' as their location.
 Following: 432, Followers: 25.

Dave Renn lists 'None' as their location.
 Following: 98, Followers: 10.

Lionel lists 'None' as their location.
 Following: 200, Followers: 200.

Megan Randall lists 'None' as their location.
 Following: 142, Followers: 98.

Jacob Salzman lists 'None' as their location.
 Following: 563, Followers: 136.

twiter not fun lists 'None' as their location.
 Following: 218, Followers: 20.

Christian Tinsley lists 'None' as th

Let's find the person who follows this handle who has the most followers. 

In [8]:
max_followers = 0

for idx, user in enumerate(followers.data) :
    followers_count = user.public_metrics['followers_count']
    
    if followers_count > max_followers :
        max_followers = followers_count
        max_follower_user = user

        
print(max_follower_user)
print(max_follower_user.public_metrics)

SpaceConscious
{'followers_count': 37561, 'following_count': 37865, 'tweet_count': 13956, 'listed_count': 305}


Let's pull some more user fields and take a look at them. The fields can be specified in the `user_fields` argument. 

In [9]:
response = client.get_user(id=user_obj.data.id,
                          user_fields=["created_at","description","location",
                                       "entities","name","pinned_tweet_id","profile_image_url",
                                       "verified","public_metrics"])

In [10]:
for field, value in response.data.items() :
    print(f"for {field} we have {value}")

for created_at we have 2009-04-18 22:08:22+00:00
for username we have 37chandler
for profile_image_url we have https://pbs.twimg.com/profile_images/2680483898/b30ae76f909352dbae5e371fb1c27454_normal.png
for description we have He/Him. Data scientist, urban cyclist, educator, erstwhile frisbee player. 

¯\_(ツ)_/¯
for public_metrics we have {'followers_count': 185, 'following_count': 592, 'tweet_count': 1048, 'listed_count': 3}
for location we have MN
for id we have 33029025
for name we have John Chandler
for verified we have False


Now a few questions for you about the user object.

Q: How many fields are being returned in the `response` object? 

A: There are 4 main fields in the response odject, data, includes, errors, and meta. There are more fields inside of these the data field for John. Data has nine fields, there could be more but nine we only requested.

---

Q: Are any of the fields within the user object non-scalar? (I.e., more complicated than a simple data type like integer, float, string, boolean, etc.) 

A: public_metrics appears to be yet another dictionary object.

---

Q: How many friends, followers, and tweets does this user have? 

A: This user has  592 friends or people he is following, 184 followers or people who is following him,and 1049 tweets


Although you won't need it for this assignment, individual tweets can be a rich source of text-based data. To illustrate the concepts, let's look at the last few tweets for this user. You are encouraged to explore the fields that are available about Tweets.

In [11]:

response = client.get_users_tweets(user_obj.data.id)

# By default, only the ID and text fields of each Tweet will be returned
for idx, tweet in enumerate(response.data) :
    print(tweet.id)
    print(tweet.text)
    print()
    
    if idx > 10 :
        break

1611545485029810180
Happy Dia de los Reyes to all who celebrate it. https://t.co/4G7zAuwC70

1608230093071212544
RT @year_progress: ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ 99%

1606038920604499969
RT @CoachBalto: This video is perfect. Parents please watch. Part 1/2 https://t.co/NvcBFmyFPO

1602407567036190743
RT @LindsayMasland: I had the realization that "grades are pretend" the first time I taught (as a TA).  

I was grading something with both…

1598645130075856896
RT @marinaendicott: My new favourite lawyer’s letter, just for the sheer joy of the tone.

1598156055997222912
RT @_TanHo: Hey friends, #AdventOfCode starts TONIGHT! I've organized a friendly leaderboard every year for the #rstats (and friends) commu…

1597746144108740608
If you like biking and not getting hit by muederboxes, you should consider one of these. https://t.co/0prMLbvj3b

1597734124927995904
RT @CraigTheDev: A lot of people argue that AI art isn't theft as it isn't copying the original images but referencing them like a person.…

15

## Pulling Follower Information

In this next section of the assignment, we will pull information about the followers of your two artists. We've seen above how to pull a set of followers using `client.get_users_followers`. This function has a parameter, `max_results`, that we can use to change the number of followers that we pull. Unfortunately, we can only pull 1000 followers at a time, which means we will need to handle the _pagination_ of our results. 

The return object has the `.data` field, where the results will be found. It also has `.meta`, which we use to select the next "page" in the results using the `next_token` result. I will illustrate the ideas using our user from above. 


### Rate Limiting

Twitter limits the rates at which we can pull data, as detailed in [this guide](https://developer.twitter.com/en/docs/twitter-api/rate-limits). We can make 15 user requests per 15 minutes, meaning that we can pull $4 \cdot 15 \cdot 1000 = 60000$ users per hour. I illustrate the handling of rate limiting below, though whether or not you hit that part of the code depends on your value of `handle`.  


In the below example, I'll pull all the followers, 25 at a time. (We're using 25 to illustrate the idea; when you do this set the value to 1000.) 

In [12]:
handle_followers = []
pulls = 0
max_pulls = 100
next_token = None

while True :

    followers = client.get_users_followers(
        user_obj.data.id, 
        max_results=25, # when you do this for real, set this to 1000!
        pagination_token = next_token,
        user_fields=["created_at","description","location",
                     "entities","name","pinned_tweet_id","profile_image_url",
                     "verified","public_metrics"]
    )
    pulls += 1
    
    for follower in followers.data : 
        follower_row = (follower.id,follower.name,follower.created_at,follower.description)
        handle_followers.append(follower_row)
    
    if 'next_token' in followers.meta and pulls < max_pulls :
        next_token = followers.meta['next_token']
    else : 
        break
        



## Pulling Twitter Data for Your Artists

Now let's take a look at your artists and see how long it is going to take to pull all their followers. 

In [13]:
artists = dict()

for handle in ['thebeatles', 'tenaciousd']:
    user_obj = client.get_user(username=handle, user_fields=["public_metrics"])
    artists[handle] = (user_obj.data.id,
                       handle,
                       user_obj.data.public_metrics['followers_count'])


for artist, data in artists.items():
    print(
        f"It would take {data[2]/(1000*15*4):.2f} hours to pull all {data[2]} followers for {artist}. ")


It would take 64.67 hours to pull all 3880389 followers for thebeatles. 
It would take 3.58 hours to pull all 214805 followers for tenaciousd. 


Depending on what you see in the display above, you may want to limit how many followers you pull. It'd be great to get at least 200,000 per artist. 

As we pull data for each artist we will write their data to a folder called "twitter", so we will make that folder if needed.

In [14]:
# Make the "twitter" folder here. If you'd like to practice your programming, add functionality
# that checks to see if the folder exists. If it does, then "unlink" it. Then create a new one.

# if not os.path.isdir("twitter"):
#     # shutil.rmtree("twitter/")
#     os.mkdir("twitter")


# You should change 'test' to your preferred folder.
MYDIR = ("twitter")
CHECK_FOLDER = os.path.isdir(MYDIR)

# If folder doesn't exist, then create it.
if not CHECK_FOLDER:
    os.makedirs(MYDIR)
    print("created folder : ", MYDIR)

else:
    print(MYDIR, "folder already exists.")

twitter folder already exists.


In this following cells, build on the above code to pull some of the followers and their data for your two artists. As you pull the data, write the follower ids to a file called `[artist name]_followers.txt` in the "twitter" folder. For instance, for Cher I would create a file named `cher_followers.txt`. As you pull the data, also store it in an object like a list or a data frame.

In addition to creating a file that only has follower IDs in it, you will create a file that includes user data. From the response object please extract and store the following fields: 

* screen_name	
* name	
* id	
* location	
* followers_count	
* friends_count	
* description

Store the fields with one user per row in a tab-delimited text file with the name `[artist name]_follower_data.txt`. For instance, for Cher I would create a file named `cher_follower_data.txt`. 

One note: the user's description can have tabs or returns in it, so make sure to clean those out of the description before writing them to the file. I've included some example code to do that below the stub. 

In [49]:
# feel free to use this to limit the number of followers you pull.
num_followers_to_pull = 200*1000 
num_followers_to_pull

200000

In [255]:
# Modify the below code stub to pull the follower IDs and write them to a file. 

handles = ['thebeatles','tenaciousd']

whitespace_pattern = re.compile(r"\s+")

user_data = dict()
followers_data = dict()

for handle in handles :
    user_data[handle] = [] # will be a list of lists
    followers_data[handle] = [] # will be a simple list of IDs

num_followers_to_pull = 2000
# Grabs the time when we start making requests to the API
start_time = datetime.datetime.now()
intervals = 60

for handle in handles :
    
    # Create the output file names 
    
    followers_output_file = handle + "_followers.txt"
    user_data_output_file = handle + "_followers_data.txt"
    
    names = ['username',
             'name',
             'id',
             'location',
             'followers_count',
             'friends_count',
             'description']
    with open('twitter\\' + user_data_output_file, 'w', encoding="utf-8") as u:
                u.write('\t'.join(map(str,names[:])) + '\n')
                
    with open('twitter\\' + followers_output_file, 'w', encoding="utf-8") as f:
                f.write(f'id' + '\n')
    
    # Using tweepy.Paginator (https://docs.tweepy.org/en/latest/v2_pagination.html), 
    # use `get_users_followers` to pull the follower data requested. 

    counter = 0
    # Instantiate a new Paginator with the Tweepy method and its arguments
    for followers in tweepy.Paginator(client.get_users_following, 
                                 id=user_obj.data.id, 
                                 max_results=100, limit =2000,
                                 user_fields=["username","name","id","location",
                                              "public_metrics","description"
                                              ]):
        for follower in followers.data: 
            clean_description = re.sub(whitespace_pattern," ",follower.description).strip()
            # Lets collect the user data.
            follower_row = [follower.username,
                            follower.name,
                            follower.id,
                            follower.location,
                            follower.public_metrics['followers_count'],
                            follower.public_metrics['following_count'],
                            clean_description]
            with open('twitter\\' + user_data_output_file, 'a', encoding="utf-8") as u:
                u.write('\t'.join(map(str,follower_row[:])) + '\n')
            # Now let's collect the user ids as a file.
            follower_id = follower.id
            with open('twitter\\' + followers_output_file, 'a', encoding="utf-8") as f:
                f.write(f'{follower_id}' + '\n')
        counter += len(followers.data)
        print(f"{handle}:", counter)
        if counter >= num_followers_to_pull:
            break # Break Here
        time.sleep(intervals)
    
        
# Let's see how long it took to grab all follower IDs
end_time = datetime.datetime.now()
print(end_time - start_time)


thebeatles: 61
tenaciousd: 61
0:02:00.609665


In [98]:
tricky_description = """
    Home by Warsan Shire
    
    no one leaves home unless
    home is the mouth of a shark.
    you only run for the border
    when you see the whole city
    running as well.

"""
# This won't work in a tab-delimited text file.
whitespace_pattern = re.compile(r"\s+")
clean_description = re.sub(whitespace_pattern," ",tricky_description).strip()
# clean_description = re.sub(r"\s+"," ",tricky_description).strip()
clean_description

'Home by Warsan Shire no one leaves home unless home is the mouth of a shark. you only run for the border when you see the whole city running as well.'

---

# Lyrics Scrape

This section asks you to pull data from the Twitter API and scrape www.AZLyrics.com. In the notebooks where you do that work you are asked to store the data in specific ways. 

In [99]:
artists = {'tenaciousd':"https://www.azlyrics.com/t/tenaciousd.html",
           'beatles':"https://www.azlyrics.com/b/beatles.html"} 
# we'll use this dictionary to hold both the artist name and the link on AZlyrics

## A Note on Rate Limiting

The lyrics site, www.azlyrics.com, does not have an explicit maximum on number of requests in any one time, but in our testing it appears that too many requests in too short a time will cause the site to stop returning lyrics pages. (Entertainingly, the page that gets returned seems to only have the song title to [a Tom Jones song](https://www.azlyrics.com/lyrics/tomjones/itsnotunusual.html).) 

Whenever you call `requests.get` to retrieve a page, put a `time.sleep(5 + 10*random.random())` on the next line. This will help you not to get blocked. If you _do_ get blocked, which you can identify if the returned pages are not correct, just request a lyrics page through your browser. You'll be asked to perform a CAPTCHA and then your requests should start working again. 

## Part 1: Finding Links to Songs Lyrics

That general artist page has a list of all songs for that artist with links to the individual song pages. 

Q: Take a look at the `robots.txt` page on www.azlyrics.com. (You can read more about these pages [here](https://developers.google.com/search/docs/advanced/robots/intro).) Is the scraping we are about to do allowed or disallowed by this page? How do you know? 

A: According to thier robots.txt, see below. We are allowed to crawl anything except the song and lyricsdb section of thier sight. 

```
User-agent: *
Disallow: /lyricsdb/
Disallow: /song/
Allow: /

User-agent: 008
Disallow: /
```


In [100]:
import random

# Let's set up a dictionary of lists to hold our links
lyrics_pages = defaultdict(list)

for artist, artist_page in artists.items() :
    # request the page and sleep
    r = requests.get(artist_page)
    time.sleep(5 + 10*random.random())

    # now extract the links to lyrics pages from this page
    # store the links `lyrics_pages` where the key is the artist and the
    # value is a list of links. 
    soup = BeautifulSoup(r.text)
    data = soup.find_all('div',attrs={'class':'listalbum-item'})
    for div in data:
        links = div.find_all('a')
        for a in links:
            lyrics_pages[artist].append("https://www.azlyrics.com" + a['href'])
    

Let's make sure we have enough lyrics pages to scrape. 

In [102]:
for artist, lp in lyrics_pages.items() :
    assert(len(set(lp)) > 20) 

In [103]:
# Let's see how long it's going to take to pull these lyrics 
# if we're waiting `5 + 10*random.random()` seconds 
for artist, links in lyrics_pages.items() : 
    print(f"For {artist} we have {len(links)}.")
    print(f"The full pull will take for this artist will take {round(len(links)*10/3600,2)} hours.")

For tenaciousd we have 104.
The full pull will take for this artist will take 0.29 hours.
For beatles we have 427.
The full pull will take for this artist will take 1.19 hours.


## Part 2: Pulling Lyrics

Now that we have the links to our lyrics pages, let's go scrape them! Here are the steps for this part. 

1. Create an empty folder in our repo called "lyrics". 
1. Iterate over the artists in `lyrics_pages`. 
1. Create a subfolder in lyrics with the artist's name. For instance, if the artist was Cher you'd have `lyrics/cher/` in your repo.
1. Iterate over the pages. 
1. Request the page and extract the lyrics from the returned HTML file using BeautifulSoup.
1. Use the function below, `generate_filename_from_url`, to create a filename based on the lyrics page, then write the lyrics to a text file with that name. 


In [168]:
def generate_filename_from_link(link) :
    
    if not link :
        return None
    
    # drop the http or https and the html
    name = link.replace("https","").replace("http","")
    name = link.replace(".html","")

    name = name.replace("/lyrics/","")
    
    # Replace useless chareacters with UNDERSCORE
    name = name.replace("://","").replace(".","_").replace("/","_")
    
    # tack on .txt
    name = name + ".txt"
    
    return(name)


In [106]:
# Make the lyrics folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then use shutil.rmtree to remove it and create a new one.

# if os.path.isdir("lyrics") : 
#     shutil.rmtree("lyrics/")

# os.mkdir("lyrics")

# You should change 'test' to your preferred folder.
MYDIR = ("lyrics")
CHECK_FOLDER = os.path.isdir(MYDIR)

# If folder doesn't exist, then create it.
if not CHECK_FOLDER:
    os.makedirs(MYDIR)
    print("created folder : ", MYDIR)

else:
    print(MYDIR, "folder already exists.")

lyrics folder already exists.


In [169]:
url_stub = "https://www.azlyrics.com" 
start = time.time()

for artist in lyrics_pages :
    total_pages = 0 
    # Use this space to carry out the following steps: 
    
    # 1. Build a subfolder for the artist
    MYDIR = ("lyrics\\" + artist)
    CHECK_FOLDER = os.path.isdir(MYDIR)

    # If folder doesn't exist, then create it.
    if not CHECK_FOLDER:
        os.makedirs(MYDIR)
        print("created folder : ", MYDIR)
    else:
        print(MYDIR, "folder already exists.")
    # 2. Iterate over the lyrics pages
    for page in lyrics_pages[artist]:
        # 3. Request the lyrics page. 
        # request the page and sleep
        r = requests.get(page)
        soup = BeautifulSoup(r.text)
        # Don't forget to add a line like `time.sleep(5 + 10*random.random())`
        # to sleep after making the request
        time.sleep(5 + 10*random.random())
        # 4. Extract the title and lyrics from the page.
        # singer = soup.find('div' , class_ = 'lyricsh').h2.text
        title = soup.find('div' , class_ = 'col-xs-12 col-lg-8 text-center').find_all('div',class_= 'div-share')[1].text.split('"')[1]
        lyrics = soup.find('div' , class_ = 'col-xs-12 col-lg-8 text-center').find_all('div')[5].text
        filename = generate_filename_from_link(page)
        with open(MYDIR + "\\" + filename, 'a', encoding="utf-8") as f:
            f.write(f'{title}' + '\n' + f"{lyrics}")
        total_pages += 1
        if total_pages >= 20:
            break # Break Here

        
    
    # 5. Write out the title, two returns ('\n'), and the lyrics. Use `generate_filename_from_url`
    #    to generate the filename. 
    
    # Remember to pull at least 20 songs per artist. It may be fun to pull all the songs for the artist
    

lyrics\tenaciousd folder already exists.
lyrics\beatles folder already exists.


In [170]:
print(f"Total run time was {round((time.time() - start)/3600,2)} hours.")

Total run time was 0.13 hours.


---

# Evaluation

This assignment asks you to pull data from the Twitter API and scrape www.AZLyrics.com.  After you have finished the above sections , run all the cells in this notebook. Print this to PDF and submit it, per the instructions.

In [171]:
# Simple word extractor from Peter Norvig: https://norvig.com/spell-correct.html
def words(text): 
    return re.findall(r'\w+', text.lower())

---

## Checking Twitter Data

The output from your Twitter API pull should be two files per artist, stored in files with formats like `cher_followers.txt` (a list of all follower IDs you pulled) and `cher_followers_data.txt`. These files should be in a folder named `twitter` within the repository directory. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [172]:
twitter_files = os.listdir("twitter")
twitter_files = [f for f in twitter_files if f != ".DS_Store"]
artist_handles = list(set([name.split("_")[0] for name in twitter_files]))

print(f"We see two artist handles: {artist_handles[0]} and {artist_handles[1]}.")

We see two artist handles: thebeatles and tenaciousd.


In [257]:
for artist in artist_handles :
    follower_file = artist + "_followers.txt"
    follower_data_file = artist + "_followers_data.txt"
    
    ids = open("twitter/" + follower_file,'r', encoding="utf-8").readlines()
    
    print(f"We see {len(ids)-1} in your follower file for {artist}, assuming a header row.")
    
    with open("twitter/" + follower_data_file,'r', encoding="utf-8") as infile :
        
        # check the headers
        headers = infile.readline().split("\t")
        
        print(f"In the follower data file ({follower_data_file}) for {artist}, we have these columns:")
        print(" : ".join(headers))
        
        description_words = []
        locations = set()
        
        
        for idx, line in enumerate(infile.readlines()) :
            line = line.strip("\n").split("\t")
            
            try : 
                locations.add(line[3])            
                description_words.extend(words(line[6]))
            except :
                pass
    
        

        print(f"We have {idx+1} data rows for {artist} in the follower data file.")

        print(f"For {artist} we have {len(locations)} unique locations.")

        print(f"For {artist} we have {len(description_words)} words in the descriptions.")
        print("Here are the five most common words:")
        print(Counter(description_words).most_common(5))

        
        print("")
        print("-"*40)
        print("")
    

We see 61 in your follower file for thebeatles, assuming a header row.
In the follower data file (thebeatles_followers_data.txt) for thebeatles, we have these columns:
username : name : id : location : followers_count : friends_count : description

We have 61 data rows for thebeatles in the follower data file.
For thebeatles we have 40 unique locations.
For thebeatles we have 663 words in the descriptions.
Here are the five most common words:
[('the', 20), ('https', 16), ('t', 16), ('co', 16), ('of', 16)]

----------------------------------------

We see 61 in your follower file for tenaciousd, assuming a header row.
In the follower data file (tenaciousd_followers_data.txt) for tenaciousd, we have these columns:
username : name : id : location : followers_count : friends_count : description

We have 61 data rows for tenaciousd in the follower data file.
For tenaciousd we have 40 unique locations.
For tenaciousd we have 663 words in the descriptions.
Here are the five most common words:

## Checking Lyrics 

The output from your lyrics scrape should be stored in files located in this path from the directory:
`/lyrics/[Artist Name]/[filename from URL]`. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [212]:
artist_folders = os.listdir("lyrics/")
artist_folders = [f for f in artist_folders if os.path.isdir("lyrics/" + f)]

for artist in artist_folders : 
    artist_files = os.listdir("lyrics/" + artist)
    artist_files = [f for f in artist_files if 'txt' in f or 'csv' in f or 'tsv' in f]

    print(f"For {artist} we have {len(artist_files)} files.")

    artist_words = []

    for f_name in artist_files : 
        with open("lyrics/" + artist + "/" + f_name) as infile : 
            artist_words.extend(words(infile.read()))

            
    print(f"For {artist} we have roughly {len(artist_words)} words, {len(set(artist_words))} are unique.")


For beatles we have 20 files.
For beatles we have roughly 3616 words, 384 are unique.
For tenaciousd we have 20 files.
For tenaciousd we have roughly 6868 words, 924 are unique.
